**Name:** Phani Valasa <br />
**UNI  :** PKV2103 <br />
**Sub.:** Personalization Course; Homework 1; Part2

**For this exercise we will try to predict the subjective quality of Portuguese red wine using only the physiochemical properties of each wine. We will not use any the common regression methods and packages for Python or R - instead we will use minimization methods in order to minimize the error of a linear model. In this case, we will define error as Residual Sum of Squares (RSS), which is also often known as Squared Error.**

In [26]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
from scipy.optimize import minimize
from collections import OrderedDict


# Read the input training set
data = pd.read_csv('winequality-red.csv',sep=';')
X = data[data.columns[:-1]]
y = data[['quality']]

# Now split the dataset into training and test data and convert them to arrays
X_train, X_test, y_train, y_test = train_test_split(X, y,test_size=0.2)

X_train = X_train.values
y_train = y_train.values
X_test = X_test.values
y_test = y_test.values

In [27]:
def compute_y(X, beta):
    y_hat = np.dot(X, beta)
    return y_hat.reshape(y_hat.shape[0],1)

def RSS(beta,X, y):
    y_hat = compute_y(X, beta)
    rss = np.sum((y-y_hat)**2)
    return rss

#### Calculation of RSS on Training and Test Data.

In [28]:
beta0 = np.random.normal(0, 1, X_train.shape[1])
res = minimize(fun=RSS, x0=beta0, args=(X_train,y_train))
beta_hat = res.x

train_RSS = RSS(beta_hat,X_train,y_train)
print "Residual sum of squares on Training data is " , round(train_RSS,2)
test_RSS = RSS(beta_hat,X_test,y_test)
print "Residual sum of squares on Test data is ", round(test_RSS,2)

Residual sum of squares on Training data is  550.97
Residual sum of squares on Test data is  117.15


#### Feature Importance and Magnitude of features

In [29]:
column_list = X.columns.values
for i in range(len(list(column_list))) :
    print column_list[i],":  ", round(beta_hat[i],4), ";   Average: ", X[column_list[i]].mean()

fixed acidity :   -0.0152 ;   Average:  8.3196372733
volatile acidity :   -1.1808 ;   Average:  0.527820512821
citric acid :   -0.1422 ;   Average:  0.270975609756
residual sugar :   0.0092 ;   Average:  2.53880550344
chlorides :   -2.0521 ;   Average:  0.0874665415885
free sulfur dioxide :   0.0036 ;   Average:  15.8749218261
total sulfur dioxide :   -0.0031 ;   Average:  46.4677923702
density :   5.1195 ;   Average:  0.996746679174
pH :   -0.6295 ;   Average:  3.31111319575
sulphates :   0.8344 ;   Average:  0.658148843027
alcohol :   0.2972 ;   Average:  10.4229831144


#### RSS for different values of beta

In [30]:
beta1=list(range(4,15))
beta2=list(range(51,40,-1))
beta3=list(range(1100,1111))

res1 = minimize(fun=RSS, x0=beta1, args=(X_train,y_train))
beta_hat1 = res1.x
test_RSS1 = RSS(beta_hat1,X_test,y_test)

res2 = minimize(fun=RSS, x0=beta2, args=(X_train,y_train))
beta_hat2 = res2.x
test_RSS2 = RSS(beta_hat2,X_test,y_test)

res3 = minimize(fun=RSS, x0=beta3, args=(X_train,y_train))
beta_hat3 = res3.x
test_RSS3 = RSS(beta_hat3,X_test,y_test)

print "RSS for different values of beta: ",round(test_RSS1,2), round(test_RSS2,2), round(test_RSS3,2)


RSS for different values of beta:  117.15 117.15 117.15


#### RSS for different Solver Methods

In [31]:
def RSS_Solver(beta, X_train, y_train, X_test, y_test, solver) :
    res = minimize(fun=RSS, x0=beta, method=solver, args=(X_train,y_train))
    beta_hat = res.x
    test_RSS = RSS(beta_hat,X_test,y_test)
    return test_RSS
    
solver_methods = ['Nelder-Mead', 'Powell', 'CG', 'BFGS', 'L-BFGS-B', 'TNC', 'COBYLA', 'SLSQP']

for i in range(len(solver_methods)):
    rss_test = RSS_Solver(beta0, X_train, y_train, X_test, y_test, solver_methods[i])
    print "Solver Method :", solver_methods[i], "; RSS Value:" , round(rss_test,2)

Solver Method : Nelder-Mead ; RSS Value: 162.38
Solver Method : Powell ; RSS Value: 121.31
Solver Method : CG ; RSS Value: 117.06
Solver Method : BFGS ; RSS Value: 117.15
Solver Method : L-BFGS-B ; RSS Value: 117.09
Solver Method : TNC ; RSS Value: 321.11
Solver Method : COBYLA ; RSS Value: 577.0
Solver Method : SLSQP ; RSS Value: 117.15


### Questions

 * **What are the qualitative results from your model? Which features seem to be most important? Do you think that the magnitude of the features in X may affect the results (for example, the average total sulfur dioxide across all wines is 46.47, but the average chlorides is only 0.087).**
 
  *Density has the highest coefficient value, which seems to indicate it is the most important feature. As the value of desity increase, it more likely to be of high quality. The magnitude of the features in X has a direct impact on the coefficients and so coefficients cannot be directly compared to understand feature importance. When magnitudes of features hugely vary, it'll help to do some standardizing of the data in pre-processing for coefficient comparision*


 * **How well does your model fit? You should be able to measure the goodness of fit, RSS, on both the training data and the test data, but only report the results on the test data. In Machine Learning we almost always only care about how well the model fits on data that has not been used to fit the model, because we need to use the model in the future, not the past. Therefore, we only report performance with holdout data, or test data.**
 
  *RSS on the training data is 550.97. And RSS on the test data is 117.15. RSS is a measure of the discrepancy between the data and an estimation model. A small RSS indicates a tight fit of the model to the data. It is used as an optimality criterion in parameter selection and model selection*
  
  
 * **Does the end result or RSS change if you try different initial values of β? What happens if you change the magnitude of the initial β?**
 
  *The end result of RSS has not changed for the different values of β used. However as the magnitude of β increase, it takes longer to converge at the optimal β*


 * **Does the choice of solver method change the end result or RSS?**
 
  *Solver Methods does change the end result of RSS. In this particular case, CG has an optimal value.*
   







### Regularizing the Model; lambda = 0.01

In [32]:
def RSS_reg_l2(beta,X, y,lamda):
    y_hat = compute_y(X, beta)
    rss_reg_l2 = np.sum((y - y_hat) ** 2) + (lamda * np.sum(beta**2))
    return rss_reg_l2

def RSS_reg_l1(beta,X, y,lamda):
    y_hat = compute_y(X, beta)
    rss_reg_l1 = np.sum((y - y_hat) ** 2) + (lamda * np.sum(np.abs(beta)))
    return rss_reg_l1

def RSS_reg(beta, X_train, y_train, X_test, y_test, reg,lamda) :
    if reg=="l2":
        res = minimize(fun=RSS_reg_l2, x0=beta, args=(X_train,y_train, lamda))
        beta_hat_l2 = res.x
        test_RSS = RSS(beta_hat_l2,X_test,y_test)
    elif reg=="l1":
        res = minimize(fun=RSS_reg_l1, x0=beta, args=(X_train,y_train,lamda))
        beta_hat_l1 = res.x
        test_RSS = RSS(beta_hat_l1,X_test,y_test)            
    return test_RSS

train_RSS_Reg_L2 = RSS_reg(beta0, X_train, y_train, X_train, y_train, "l2",0.01)
test_RSS_Reg_L2 = RSS_reg(beta0, X_train, y_train, X_test, y_test, "l2",0.01)

print "Regularized Model: L2 Residual sum of squares on Training data is ", round(train_RSS_Reg_L2,2)
print "Regularized Model: L2 Residual sum of squares on Test data is ", round(test_RSS_Reg_L2,2)

Regularized Model: L2 Residual sum of squares on Training data is  550.98
Regularized Model: L2 Residual sum of squares on Test data is  117.07


#### Different values of Lamda & Training RSS; L2 Regularization

In [33]:
lamda_reg = [0,0.01, 0.02, 0.03, 0.04, 0.05, 0.06, 0.07, 0.08, 0.09, 0.1, 0.2,0.3,0.4,0.5,1,10,20,30,40,50]

for i in range(len(lamda_reg)):
    train_RSS_Reg_L2 = 0
    train_RSS_Reg_L2 = RSS_reg(beta0, X_train, y_train, X_train, y_train , "l2", lamda_reg[i])  
    print "Lambda : ", lamda_reg[i], "; L2 RSS (Train): ", round(train_RSS_Reg_L2,4)

Lambda :  0 ; L2 RSS (Train):  550.973
Lambda :  0.01 ; L2 RSS (Train):  550.9768
Lambda :  0.02 ; L2 RSS (Train):  550.9878
Lambda :  0.03 ; L2 RSS (Train):  551.0055
Lambda :  0.04 ; L2 RSS (Train):  551.0294
Lambda :  0.05 ; L2 RSS (Train):  551.0589
Lambda :  0.06 ; L2 RSS (Train):  551.0938
Lambda :  0.07 ; L2 RSS (Train):  551.1336
Lambda :  0.08 ; L2 RSS (Train):  551.1779
Lambda :  0.09 ; L2 RSS (Train):  551.2263
Lambda :  0.1 ; L2 RSS (Train):  551.2787
Lambda :  0.2 ; L2 RSS (Train):  551.9583
Lambda :  0.3 ; L2 RSS (Train):  552.7979
Lambda :  0.4 ; L2 RSS (Train):  553.6912
Lambda :  0.5 ; L2 RSS (Train):  554.5842
Lambda :  1 ; L2 RSS (Train):  558.4808
Lambda :  10 ; L2 RSS (Train):  576.3298
Lambda :  20 ; L2 RSS (Train):  584.7542
Lambda :  30 ; L2 RSS (Train):  591.5533
Lambda :  40 ; L2 RSS (Train):  597.29
Lambda :  50 ; L2 RSS (Train):  602.1947


#### Different Values of Lamda & Test RSS ; L2 Regularization

In [34]:
lamda_reg = [0,0.01, 0.02, 0.03, 0.04, 0.05, 0.06, 0.07, 0.08, 0.09, 0.1, 0.2,0.3,0.4,0.5,1,10,20,30,40,50]

for i in range(len(lamda_reg)):
    test_RSS_Reg_L2 = 0
    test_RSS_Reg_L2 = RSS_reg(beta0, X_train, y_train, X_test, y_test , "l2", lamda_reg[i])  
    print "Lambda : ", lamda_reg[i], "; L2 RSS (Test): ", round(test_RSS_Reg_L2,3)

Lambda :  0 ; L2 RSS (Test):  117.149
Lambda :  0.01 ; L2 RSS (Test):  117.066
Lambda :  0.02 ; L2 RSS (Test):  116.987
Lambda :  0.03 ; L2 RSS (Test):  116.911
Lambda :  0.04 ; L2 RSS (Test):  116.839
Lambda :  0.05 ; L2 RSS (Test):  116.771
Lambda :  0.06 ; L2 RSS (Test):  116.705
Lambda :  0.07 ; L2 RSS (Test):  116.643
Lambda :  0.08 ; L2 RSS (Test):  116.583
Lambda :  0.09 ; L2 RSS (Test):  116.526
Lambda :  0.1 ; L2 RSS (Test):  116.471
Lambda :  0.2 ; L2 RSS (Test):  116.037
Lambda :  0.3 ; L2 RSS (Test):  115.75
Lambda :  0.4 ; L2 RSS (Test):  115.556
Lambda :  0.5 ; L2 RSS (Test):  115.422
Lambda :  1 ; L2 RSS (Test):  115.167
Lambda :  10 ; L2 RSS (Test):  115.257
Lambda :  20 ; L2 RSS (Test):  115.657
Lambda :  30 ; L2 RSS (Test):  116.192
Lambda :  40 ; L2 RSS (Test):  116.737
Lambda :  50 ; L2 RSS (Test):  117.252


#### L1 Regularization

In [38]:
lamda_reg = [0, 0.01, 0.02, 0.03, 0.04, 0.05, 0.06, 0.07, 0.08, 0.09, 0.1, 0.2,0.3,0.4,0.5,1,10,20,30,40,50]

for i in range(len(lamda_reg)):
    train_RSS_Reg_L1 = RSS_reg(beta0, X_train, y_train, X_train, y_train, "l1", lamda_reg[i])  
    print "Lambda : ", lamda_reg[i], "; L1 RSS (Train): ", round(train_RSS_Reg_L1,4)

Lambda :  0 ; L1 RSS (Train):  550.973
Lambda :  0.01 ; L1 RSS (Train):  550.9731
Lambda :  0.02 ; L1 RSS (Train):  550.9733
Lambda :  0.03 ; L1 RSS (Train):  550.9736
Lambda :  0.04 ; L1 RSS (Train):  550.9741
Lambda :  0.05 ; L1 RSS (Train):  550.9748
Lambda :  0.06 ; L1 RSS (Train):  550.9756
Lambda :  0.07 ; L1 RSS (Train):  550.9765
Lambda :  0.08 ; L1 RSS (Train):  550.9776
Lambda :  0.09 ; L1 RSS (Train):  550.9788
Lambda :  0.1 ; L1 RSS (Train):  550.9801
Lambda :  0.2 ; L1 RSS (Train):  551.0015
Lambda :  0.3 ; L1 RSS (Train):  551.0371
Lambda :  0.4 ; L1 RSS (Train):  551.087
Lambda :  0.5 ; L1 RSS (Train):  551.1512
Lambda :  1 ; L1 RSS (Train):  551.6859
Lambda :  10 ; L1 RSS (Train):  577.3858
Lambda :  20 ; L1 RSS (Train):  584.3097
Lambda :  30 ; L1 RSS (Train):  591.7746
Lambda :  40 ; L1 RSS (Train):  590.9881
Lambda :  50 ; L1 RSS (Train):  615.6613


In [39]:
lamda_reg = [0, 0.01, 0.02, 0.03, 0.04, 0.05, 0.06, 0.07, 0.08, 0.09, 0.1, 0.2,0.3,0.4,0.5,1,10,20,30,40,50]

for i in range(len(lamda_reg)):
    test_RSS_Reg_L1 = RSS_reg(beta0, X_train, y_train, X_test, y_test, "l1", lamda_reg[i])  
    print "Lambda : ", lamda_reg[i], "; L1 RSS (Test): ", round(test_RSS_Reg_L1,4)

Lambda :  0 ; L1 RSS (Test):  117.1485
Lambda :  0.01 ; L1 RSS (Test):  117.1381
Lambda :  0.02 ; L1 RSS (Test):  117.1277
Lambda :  0.03 ; L1 RSS (Test):  117.1174
Lambda :  0.04 ; L1 RSS (Test):  117.1071
Lambda :  0.05 ; L1 RSS (Test):  117.0968
Lambda :  0.06 ; L1 RSS (Test):  117.0865
Lambda :  0.07 ; L1 RSS (Test):  117.0763
Lambda :  0.08 ; L1 RSS (Test):  117.0661
Lambda :  0.09 ; L1 RSS (Test):  117.056
Lambda :  0.1 ; L1 RSS (Test):  117.0458
Lambda :  0.2 ; L1 RSS (Test):  116.9462
Lambda :  0.3 ; L1 RSS (Test):  116.8497
Lambda :  0.4 ; L1 RSS (Test):  116.7562
Lambda :  0.5 ; L1 RSS (Test):  116.6659
Lambda :  1 ; L1 RSS (Test):  116.2605
Lambda :  10 ; L1 RSS (Test):  116.0598
Lambda :  20 ; L1 RSS (Test):  115.7155
Lambda :  30 ; L1 RSS (Test):  116.7202
Lambda :  40 ; L1 RSS (Test):  117.8729
Lambda :  50 ; L1 RSS (Test):  119.6407


### Questions

 * **Try adding in an L2 (aka Ridge) regularization penalty to your model above to create a new, regularized model.  You will need to choose a value of lambda, so start with something small, like 0.01**
 
  *Regularized Model: Created the L2 model and found that at Lamda = 0.01, the Residual sum of squares on Test data is  117.04*


 * **How does RSS on the training data change? How does RSS on the test data change?**
 
  * For the above code run at Lambda - 0.01, RSS on the training data has increased marginally (i.e., from 550.973 to 550.9768) while the RSS on test data decreased a little (i.e., from 117.149 to 117.066) *
  
  
 * **What happens if you try different values of lambda? Can you roughly tune lambda to get the best results on the test data?**
 
  *When different values of Lamdba were tested, the training data RSS has increased with Lambda. Wile the Test data RSS has decresed to a point where RSS is 115.167 for Lambda = 1, and then started to increase indicating the optimum RSS is at the value of Lambda =1*


 * **Now try using an L1 (aka Lasso) regularization penalty instead. Report your findings on how RSS changes, and if you can roughly tune lambda**
 
  *With L1, RSS on test data has decreased all the way until Lambda = 20 (RSS : 115.7155), and then it started to increase again. The tuned lambda value is 20, which is where there is an optimal RSS value. Please see the output above for more details*


 * **Again, do you think that the magnitude of the features in X may affect the results with regularization?**
 
  *By introducting the regularization objective function, the beta coefficients are reduced. The magnitube of the features do effect the Coefficients. However Regularization penalizes the huge value of cooefficients to keep the overall objective function minimum. L2 regularization tends to penalize the coefficients with higher values much more than the lower value ones. L1 regularization penalized all of the coefficients equally.*
   





